### DISCUSSION TITLE
WebKit in Kas?

### DISCUSSION THREAD

**Discussion by NovaAndrom3da** [kas-gui/kas#402] [Q&A] [answered]
I'm thinking about writing a web browser using Kas- is there any way I could embed WebKit into Kas?

**Comment by dhardy** [comment, marked as answer]
Hi,

Interesting question. Not one I've put much thought into yet, to be honest.

WebKit has ports for WPE (embedded) and GTK. Most existing Rust crates appear to use WebKitGTK.

My best idea for a quick solution would be use a child window (see [`WindowBuilder::with_parent_window`](https://docs.rs/winit/latest/winit/window/struct.WindowBuilder.html#method.with_parent_window)) and run the browser in a different window. This is only supporting on X11 and Windows so far. Alternatively just to use a toolkit with better support for browsers (e.g. Tauri).

A more ambitious option could be to port Kas to [Tao](https://crates.io/crates/tao/) (replacing some Winit code and most of the drawing routines), which should make compatibility with WebKitGTK more feasible. I may do some related work in the future (looking at using Piet, which has implementations for both Cairo and WGPU).

Of course of the goal is to use GTK, you might just use GTK (or e.g. Relm4); if not, maybe it would be better to try looking at Servo.
